File: Core/src/main/java/siminov/core/database/sqlite/DatabaseImpl.java
/** 
 * [SIMINOV FRAMEWORK]
 * Copyright [2015] [Siminov Software Solution LLP|support@siminov.com]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

package siminov.core.database.sqlite;

import java.io.File;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.Map;

import siminov.core.database.DatabaseUtils;
import siminov.core.database.design.IDataTypeHandler;
import siminov.core.database.design.IDatabaseImpl;
import siminov.core.exception.DatabaseException;
import siminov.core.exception.DeploymentException;
import siminov.core.log.Log;
import siminov.core.model.DatabaseDescriptor;
import siminov.core.model.EntityDescriptor;
import siminov.core.model.EntityDescriptor.Attribute;
import android.database.sqlite.SQLiteCursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteException;
import android.database.sqlite.SQLiteStatement;


/**
 * Provides IDatabase implementation for SQLite database.
 */
public class DatabaseImpl implements IDatabaseImpl {

	private SQLiteDatabase sqliteDatabase = null;
	
	public void openOrCreate(final DatabaseDescriptor databaseDescriptor) throws DatabaseException {

		String databasePath = new DatabaseUtils().getDatabasePath(databaseDescriptor);

		File file = new File(databasePath);
		if(!file.exists()) {
			try {
				file.mkdirs();
			} catch(Exception exception) {
				Log.error(DatabaseImpl.class.getName(), "openOrCreate", "Exception caught while creating database directories, DATABASE-PATH: " + databasePath + ", DATABASE-DESCRIPTOR: " + databaseDescriptor.getDatabaseName() + ", " + exception.getMessage());
				throw new DeploymentException(DatabaseImpl.class.getName(), "openOrCreate", "Exception caught while creating database directories, DATABASE-PATH: " + databasePath + ", DATABASE-DESCRIPTOR: " + databaseDescriptor.getDatabaseName() + ", " + exception.getMessage());
			}
		}
		
		String databaseName = databaseDescriptor.getDatabaseName();
		if(databaseName == null || databaseName.length() <= 0) {
			Log.error(getClass().getName(), "doValidation", "DATABASE-NAME IS MANDATORY FIELD - DATABASE-DESCRIPTOR: " + databaseDescriptor.getDatabaseName());
			throw new DatabaseException(getClass().getName(), "doValidation", "DATABASE-NAME IS MANDATORY FIELD - DATABASE-DESCRIPTOR: " + databaseDescriptor.getDatabaseName());
		}
		
		if(!databaseName.endsWith(".db")) {
			databaseName = databaseName + ".db";
		}
		
		try {
			sqliteDatabase = SQLiteDatabase.openDatabase(databasePath + databaseName, null, SQLiteDatabase.CREATE_IF_NECESSARY);
		} catch(SQLiteException sqliteException) {
			Log.error(DatabaseImpl.class.getName(), "openOrCreate", "SQLiteException caught while opening database, " + sqliteException.getMessage());
			throw new DatabaseException(DatabaseImpl.class.getName(), "openOrCreate", "SQLiteException caught while opening database, " + sqliteException.getMessage());
		}
	}
	
	public void close(final DatabaseDescriptor databaseDescriptor) throws DatabaseException {
		sqliteDatabase.close();
	}

	public void executeQuery(final DatabaseDescriptor databaseDescriptor, final EntityDescriptor entityDescriptor, final String query) throws DatabaseException {
		Log.debug(DatabaseImpl.class.getName(), "executeQuery(" + query + ")", "QUERY: " + query);
		
		try {
			sqliteDatabase.execSQL(query);			
		} catch(SQLiteException sqliteException) {
			Log.error(DatabaseImpl.class.getName(), "executeQuery(" + query + ")", "SQLiteException caught while executing query, QUERY: " + query + ", " + sqliteException.getMessage());
			throw new DatabaseException(DatabaseImpl.class.getName(), "executeQuery(" + query + ")", "SQLiteException caught while executing query, QUERY: " + query + ", " + sqliteException.getMessage());
		}
	}
	
	public void executeBindQuery(final DatabaseDescriptor databaseDescriptor, final EntityDescriptor entityDescriptor, final String query, final Iterator<Object> columnValues) throws DatabaseException {
		Log.debug(DatabaseImpl.class.getName(), "executeBindQuery", "QUERY: " + query);
		
		SQLiteStatement statement =null;
		
		try {
			statement = sqliteDatabase.compileStatement(query);
		} catch(SQLiteException sqliteException) {
			Log.error(DatabaseImpl.class.getName(), "executeBindQuery(" + query + ")", "SQLiteException caught while compiling statement, " + sqliteException.getMessage());

			int index = 0;
			while(columnValues.hasNext()) {
				Object columnValue = columnValues.next();
				
				if(columnValue instanceof String) {
					Log.error(DatabaseImpl.class.getName(), "executeBindQuery", "COLUMN-INDEX: " + index + ", VALUE: " + columnValue);
				} else if(columnValue instanceof Integer) {
					Log.error(DatabaseImpl.class.getName(), "executeBindQuery", "COLUMN-INDEX " + index + ", VALUE: " + ((Integer) columnValue).intValue());
				} else if(columnValue instanceof Long) {
					Log.error(DatabaseImpl.class.getName(), "executeBindQuery", "COLUMN-INDEX " + index + ", VALUE: " + ((Long) columnValue).longValue());
				} else if(columnValue instanceof Double) {
					Log.error(DatabaseImpl.class.getName(), "executeBindQuery", "COLUMN-INDEX " + index + ", VALUE: " + ((Double) columnValue).doubleValue());
				} else if(columnValue instanceof Float) {
					Log.error(DatabaseImpl.class.getName(), "executeBindQuery", "COLUMN-INDEX " + index + ", VALUE: " + ((Float) columnValue).floatValue());
				} else if(columnValue instanceof byte[]) {
					Log.error(DatabaseImpl.class.getName(), "executeBindQuery", "COLUMN-INDEX " + index + ", VALUE: " + new String((byte[]) columnValue));
				} else if(columnValue instanceof Boolean) {
					Log.error(DatabaseImpl.class.getName(), "executeBindQuery", "COLUMN-INDEX " + index + ", VALUE: " + ((Boolean) columnValue).toString());
				} else {
					Log.error(DatabaseImpl.class.getName(), "executeBindQuery", "COLUMN-INDEX " + index + ", VALUE: " + columnValue);
				}
				
				index++;
			}

			throw new DatabaseException(DatabaseImpl.class.getName(), "executeBindQuery(" + query + ")", "SQLiteException caught while compiling statement, " + sqliteException.getMessage());
		}
		
		final Collection<Object> duplicateColumnValues = new LinkedList<Object>();
		
		int index = 0;
		while(columnValues.hasNext()) {
			Object columnValue = columnValues.next();
			
			if(columnValue instanceof String) {
				statement.bindString(index + 1, (String) columnValue);
			} else if(columnValue instanceof Integer) {
				statement.bindLong(index + 1, ((Integer) columnValue));
			} else if(columnValue instanceof Long) {
				statement.bindLong(index + 1, (Long) columnValue);
			} else if(columnValue instanceof Double) {
				statement.bindDouble(index + 1, (Double) columnValue);
			} else if(columnValue instanceof Float) {
				statement.bindDouble(index + 1, (Float) columnValue);
			} else if(columnValue instanceof byte[]) {
				statement.bindBlob(index + 1, (byte[]) columnValue);
			} else if(columnValue instanceof Boolean) {
				statement.bindString(index + 1,  (Boolean) columnValue ? Boolean.TRUE.toString() : Boolean.FALSE.toString());
			} else {
				statement.bindNull(index + 1);
			}
			
			duplicateColumnValues.add(columnValue);
			index++;
		} 

		try {
			statement.execute();
		} catch(SQLiteException sqliteException) {
			Log.error(DatabaseImpl.class.getName(), "executeBindQuery(" + query + ")", "SQLiteException caught while executing statement, " + sqliteException.getMessage());
			
			index = 0;

			Iterator<Object> newColumnValues = duplicateColumnValues.iterator();
			while(newColumnValues.hasNext()) {
				Object columnValue = newColumnValues.next();
				
				if(columnValue instanceof String) {
					Log.error(DatabaseImpl.class.getName(), "executeBindQuery", "COLUMN-NAME: " + index + ", VALUE: " + columnValue);
				} else if(columnValue instanceof Integer) {
					Log.error(DatabaseImpl.class.getName(), "executeBindQuery", "COLUMN-NAME: " + index + ", VALUE: " + ((Integer) columnValue).longValue());
				} else if(columnValue instanceof Long) {
					Log.error(DatabaseImpl.class.getName(), "executeBindQuery", "COLUMN-NAME: " + index + ", VALUE: " + ((Long) columnValue).longValue());
				} else if(columnValue instanceof Double) {
					Log.error(DatabaseImpl.class.getName(), "executeBindQuery", "COLUMN-NAME: " + index + ", VALUE: " + ((Double) columnValue).doubleValue());
				} else if(columnValue instanceof Float) {
					Log.error(DatabaseImpl.class.getName(), "executeBindQuery", "COLUMN-NAME: " + index + ", VALUE: " + ((Float) columnValue).floatValue());
				} else if(columnValue instanceof byte[]) {
					Log.error(DatabaseImpl.class.getName(), "executeBindQuery", "COLUMN-NAME: " + index + ", VALUE: " + new String((byte[]) columnValue));
				} else if(columnValue instanceof Boolean) {
					Log.error(DatabaseImpl.class.getName(), "executeBindQuery", "COLUMN-NAME: " + index + ", VALUE: " + ((Boolean) columnValue).toString());
				} else {
					Log.error(DatabaseImpl.class.getName(), "executeBindQuery", "COLUMN-INDEX " + index + ", VALUE: " + columnValue);
				}
				
				index++;
			}
			
			statement.clearBindings();
			statement.close();

			throw new DatabaseException(DatabaseImpl.class.getName(), "executeBindQuery(" + query + ")", "SQLiteException caught while executing statement, " + sqliteException.getMessage());
		}

		statement.clearBindings();
		statement.close();
	}
	
	public Iterator<Map<String, Object>> executeSelectQuery(final DatabaseDescriptor databaseDescriptor, final EntityDescriptor entityDescriptor, final String query) throws DatabaseException {
		
			SQLiteCursor sqliteCursor = (SQLiteCursor) sqliteDatabase.rawQuery(query, null);
			Collection<Map<String, Object>> tuples = new ArrayList<Map<String,Object>>();
			
			while(sqliteCursor.moveToNext()) {
				HashMap<String, Object> tuple = new HashMap<String, Object>();
				
				String[] columnNames = sqliteCursor.getColumnNames();
				if(columnNames == null || columnNames.length <= 0) {
					continue;
				}
				
				for(int i = 0;i < columnNames.length;i++) {
					boolean isString = sqliteCursor.isString(i);
					boolean isLong = sqliteCursor.isLong(i);
					boolean isFloat = sqliteCursor.isFloat(i);
					boolean isBlob = sqliteCursor.isBlob(i);
					
					if(isString) {

						String columnValue = sqliteCursor.getString(i);

						if(entityDescriptor != null) {
							
							Attribute attribute = entityDescriptor.getAttributeBasedOnColumnName(columnNames[i]);
							if(attribute != null) {
	
								if(attribute.getType().equalsIgnoreCase(IDataTypeHandler.STRING_DATA_TYPE)) {
									tuple.put(columnNames[i], (String) columnValue);
								} else if(attribute.getType().equalsIgnoreCase(IDataTypeHandler.PRIMITIVE_BOOLEAN_DATA_TYPE)) {
									tuple.put(columnNames[i], columnValue.equalsIgnoreCase(Boolean.TRUE.toString()) ? true : false);
								} else if(attribute.getType().equalsIgnoreCase(IDataTypeHandler.BOOLEAN_DATA_TYPE)) {
									tuple.put(columnNames[i], columnValue.equalsIgnoreCase(Boolean.TRUE.toString()) ? Boolean.TRUE : Boolean.FALSE);
								}
							} else {
								tuple.put(columnNames[i], columnValue);
							}
						} else {
							tuple.put(columnNames[i], columnValue);
						}
					} else if(isLong) {
						long columnValue = sqliteCursor.getLong(i);

						if(entityDescriptor != null) {
							
							Attribute attribute = entityDescriptor.getAttributeBasedOnColumnName(columnNames[i]);
							if(attribute != null) {
								if(attribute.getType().equalsIgnoreCase(IDataTypeHandler.PRIMITIVE_INTEGER_DATA_TYPE)) {
									tuple.put(columnNames[i], Integer.parseInt(Long.toString(columnValue)));
								} else if(attribute.getType().equalsIgnoreCase(IDataTypeHandler.INTEGER_DATA_TYPE)) {
									tuple.put(columnNames[i], Integer.getInteger(Long.toString(columnValue)));
								} else if(attribute.getType().equalsIgnoreCase(IDataTypeHandler.PRIMITIVE_LONG_DATA_TYPE)) {
									tuple.put(columnNames[i], columnValue);
								} else if(attribute.getType().equalsIgnoreCase(IDataTypeHandler.LONG_DATA_TYPE)) {
									tuple.put(columnNames[i], Long.valueOf(columnValue));
								}
							} else {
								tuple.put(columnNames[i], columnValue);
							}
						} else {
							tuple.put(columnNames[i], columnValue);
						}
					} else if(isFloat) {
						float columnValue = sqliteCursor.getFloat(i);
						
						if(entityDescriptor != null) {
							
							Attribute attribute = entityDescriptor.getAttributeBasedOnColumnName(columnNames[i]);
							if(attribute != null) {
								if(attribute.getType().equalsIgnoreCase(IDataTypeHandler.PRIMITIVE_DOUBLE_DATA_TYPE)) {
									tuple.put(columnNames[i], Double.parseDouble(Float.toString(columnValue)));
								} else if(attribute.getType().equalsIgnoreCase(IDataTypeHandler.DOUBLE_DATA_TYPE)) {
									tuple.put(columnNames[i], Double.valueOf(Float.toString(columnValue)));
								} else if(attribute.getType().equalsIgnoreCase(IDataTypeHandler.PRIMITIVE_FLOAT_DATA_TYPE)) {
									tuple.put(columnNames[i], columnValue);
								} else if(attribute.getType().equalsIgnoreCase(IDataTypeHandler.FLOAT_DATA_TYPE)) {
									tuple.put(columnNames[i], Float.valueOf(columnValue));
								}
							} else {
								tuple.put(columnNames[i], columnValue);
							}
						} else {
							tuple.put(columnNames[i], columnValue);
						}

						tuple.put(columnNames[i], columnValue);
					} else if(isBlob) {
						tuple.put(columnNames[i], sqliteCursor.getBlob(i));
					}
				}
				
				tuples.add(tuple);
			}
			
			sqliteCursor.close();
		return tuples.iterator();
	}
	
	public void executeMethod(final String methodName, final Object parameter) throws DatabaseException {
		Method method = null;
		
		if(parameter instanceof Boolean) {
			try {
				method = sqliteDatabase.getClass().getMethod(methodName, boolean.class);	
			} catch(Exception exception) {
				Log.error(DatabaseImpl.class.getName(), "invokeMethod", "Exception caught while getting method, METHOD-NAME: " + methodName + ", " + exception.getMessage());
				throw new DatabaseException(DatabaseImpl.class.getName(), "invokeMethod", "Exception caught while getting method, METHOD-NAME: " + methodName + ", " + exception.getMessage());
			}
		} else {
			if(parameter == null) {
				try {
					method = sqliteDatabase.getClass().getMethod(methodName, new Class[] {});	
				} catch(Exception exception) {
					Log.error(DatabaseImpl.class.getName(), "invokeMethod", "Exception caught while getting method, METHOD-NAME: " + methodName + ", " + exception.getMessage());
					throw new DatabaseException(DatabaseImpl.class.getName(), "invokeMethod", "Exception caught while getting method, METHOD-NAME: " + methodName + ", " + exception.getMessage());
				}
			} else {
				try {
					method = sqliteDatabase.getClass().getMethod(methodName, parameter.getClass());	
				} catch(Exception exception) {
					Log.error(DatabaseImpl.class.getName(), "invokeMethod", "Exception caught while getting method, METHOD-NAME: " + methodName + ", " + exception.getMessage());
					throw new DatabaseException(DatabaseImpl.class.getName(), "invokeMethod", "Exception caught while getting method, METHOD-NAME: " + methodName + ", " + exception.getMessage());
				}
			}
			
		}
		
		
		method.setAccessible(true);
		
		if(parameter == null) {
			try {
				 method.invoke(sqliteDatabase, new Object[] {});	
			} catch(Exception exception) {
				Log.error(DatabaseImpl.class.getName(), "invokeMethod", "Exception caught while getting return value from method, METHOD-NAME: " + methodName + ", " + exception.getMessage());
				throw new DatabaseException(DatabaseImpl.class.getName(), "invokeMethod", "Exception caught while getting return value from method, METHOD-NAME: " + methodName + ", " + exception.getMessage());
			}
		} else {
			try {
				 method.invoke(sqliteDatabase, new Object[] {parameter});	
			} catch(Exception exception) {
				Log.error(DatabaseImpl.class.getName(), "invokeMethod", "Exception caught while getting return value from method, METHOD-NAME: " + methodName + ", " + exception.getMessage());
				throw new DatabaseException(DatabaseImpl.class.getName(), "invokeMethod", "Exception caught while getting return value from method, METHOD-NAME: " + methodName + ", " + exception.getMessage());
			}
		}
	}

}

File: Windows/universal8.1/Siminov/Core/Database/SQLite/DatabaseImpl.cs
/** 
 * [SIMINOV FRAMEWORK]
 * Copyright [2015] [Siminov Software Solution LLP|support@siminov.com]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/


#if __MOBILE__
#define XAMARIN
#endif

#if !__MOBILE__
#define WINDOWS
#endif


using Siminov.Core.Utils;
using Siminov.Core.Database.Design;
using Siminov.Core.Exception;
using Siminov.Core.Log;
using Siminov.Core.Model;
using SQLite;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
using System.IO;

namespace Siminov.Core.Database.Sqlite
{

    /// <summary>
    /// Provides IDatabase implementation for SQLite database.
    /// </summary>
    public class DatabaseImpl : IDatabaseImpl
    {

        private SQLiteConnection sqliteDatabase = null;

        public void OpenOrCreate(DatabaseDescriptor databaseDescriptor)
        {

            String databasePath = new DatabaseUtils().GetDatabasePath(databaseDescriptor);

            if (!FileUtils.DoesFolderExists(databasePath, FileUtils.LOCAL_FOLDER))
            {
                try
                {
                    FileUtils.CreateFolder(databasePath, FileUtils.LOCAL_FOLDER);
                }
                catch (System.Exception exception)
                {
                    Log.Log.Error(typeof(DatabaseImpl).FullName, "OpenOrCreate", "Exception caught while creating database directories, DATABASE-PATH: " + databasePath + ", DATABASE-DESCRIPTOR: " + databaseDescriptor.GetDatabaseName() + ", " + exception.Message);
                    throw new DeploymentException(typeof(DatabaseImpl).FullName, "OpenOrCreate", "Exception caught while creating database directories, DATABASE-PATH: " + databasePath + ", DATABASE-DESCRIPTOR: " + databaseDescriptor.GetDatabaseName() + ", " + exception.Message);
                }
            }

            String databaseName = databaseDescriptor.GetDatabaseName();
            if (databaseName == null || databaseName.Length <= 0)
            {
                Log.Log.Error(typeof(DatabaseImpl).FullName, "OpenOrCreate", "DATABASE-NAME IS MANDATORY FIELD - DATABASE-DESCRIPTOR: " + databaseDescriptor.GetDatabaseName());
                throw new DatabaseException(typeof(DatabaseImpl).FullName, "OpenOrCreate", "DATABASE-NAME IS MANDATORY FIELD - DATABASE-DESCRIPTOR: " + databaseDescriptor.GetDatabaseName());
            }

            if (!databaseName.EndsWith(".db"))
            {
                databaseName = databaseName + ".db";
            }

            try
            {

                #if XAMARIN
                    String databaseURI = Path.Combine (databasePath, databaseName);
                    sqliteDatabase = new SQLiteConnection(databaseURI);
                #elif WINDOWS
                    sqliteDatabase = new SQLiteConnection(databasePath + databaseName);
                #endif
            }
            catch (SQLiteException sqliteException)
            {
                Log.Log.Error(typeof(DatabaseImpl).FullName, "OpenOrCreate", "SQLiteException caught while opening database, " + sqliteException.Message);
                throw new DatabaseException(typeof(DatabaseImpl).FullName, "OpenOrCreate", "SQLiteException caught while opening database, " + sqliteException.Message);
            }
        }

        public void Close(DatabaseDescriptor databaseDescriptor)
        {
            sqliteDatabase.Close();
        }

        public void ExecuteQuery(DatabaseDescriptor databaseDescriptor, EntityDescriptor entityDescriptor, String query)
        {
            Log.Log.Debug(typeof(DatabaseImpl).FullName, "ExecuteQuery(" + query + ")", "QUERY: " + query);

            try
            {
                sqliteDatabase.Execute(query);
            }
            catch (SQLiteException sqliteException)
            {
                if (sqliteDatabase.IsInTransaction)
                {
                    sqliteDatabase.Rollback();
                }

                Log.Log.Error(typeof(DatabaseImpl).FullName, "ExecuteQuery(" + query + ")", "SQLiteException caught while executing query, QUERY: " + query + ", " + sqliteException.Message);
                throw new DatabaseException(typeof(DatabaseImpl).FullName, "ExecuteQuery(" + query + ")", "SQLiteException caught while executing query, QUERY: " + query + ", " + sqliteException.Message);
            }
        }

        public void ExecuteBindQuery(DatabaseDescriptor databaseDescriptor, EntityDescriptor entityDescriptor, String query, IEnumerator<Object> columnValues)
        {
            Log.Log.Debug(typeof(DatabaseImpl).FullName, "ExecuteBindQuery", "QUERY: " + query);

            ICollection<Object> values = new List<Object>();
            while (columnValues.MoveNext())
            {
                values.Add(columnValues.Current);
            }

            try
            {
                sqliteDatabase.Execute(query, values.ToArray());
            }
            catch (SQLiteException sqliteException)
            {
                if (sqliteDatabase.IsInTransaction)
                {
                    sqliteDatabase.Rollback();
                }

                Log.Log.Error(typeof(DatabaseImpl).FullName, "ExecuteBindQuery(" + query + ")", "SQLiteException caught while executing query, QUERY: " + query + ", " + sqliteException.Message);
                throw new DatabaseException(typeof(DatabaseImpl).FullName, "ExecuteBindQuery(" + query + ")", "SQLiteException caught while executing query, QUERY: " + query + ", " + sqliteException.Message);
            }
        }

        public IEnumerator<IDictionary<String, Object>> ExecuteSelectQuery(DatabaseDescriptor databaseDescriptor, EntityDescriptor entityDescriptor, String query)
        {

            #if XAMARIN
                SQLitePCL.sqlite3_stmt statement = SQLite3.Prepare2(sqliteDatabase.Handle, query);
            #elif WINDOWS
                List<SQLiteQueryRow> statement = null;
            #endif

            try
            {

                #if XAMARIN
                    statement = SQLite3.Prepare2(sqliteDatabase.Handle, query);
                #elif WINDOWS
                    statement = sqliteDatabase.Query2(query);
                #endif
                
            }
            catch (System.Exception exception)
            {
                if (sqliteDatabase.IsInTransaction)
                {
                    sqliteDatabase.Rollback();
                }

                Log.Log.Error(typeof(Database).FullName, "ExecuteSelectQuery", "Exception caught while executing the select query, QUERY: " + query);
                throw new DatabaseException(typeof(Database).FullName, "ExecuteSelectQuery", exception.Message);
            }

            List<Dictionary<String, Object>> tuples = new List<Dictionary<String, Object>>();

            #if XAMARIN

                SQLite3.Result result;
			    while ((result = SQLite3.Step (statement)) == SQLite3.Result.Row)
			    {
				    IDictionary<String, Object> tuple = new Dictionary<String, Object>();

				    String stmtResult = result.ToString ();

				    //string[] names = SQLite3.ColType.GetNames ();
				    //string[] values = SQLite3.ColType.GetValues ();
				    int columnsCount = SQLite3.ColumnCount (statement);
				    for(int i = 0;i < columnsCount;i++) 
				    {
					    String columnName = SQLite3.ColumnName (statement, i);
					    SQLite3.ColType columnType = SQLite3.ColumnType (statement, i);
					    Object columnValue = SQLite3.ColumnText (statement, i);

					    bool isString = false;
					    bool isLong = false;
					    bool isFloat = false;
					    bool isBlob = false;

					    if (columnType == SQLite3.ColType.Text) 
					    {
						    isString = true;
					    } 
					    else if (columnType == SQLite3.ColType.Integer) 
					    {
						    isLong = true;
					    }
					    else if(columnType == SQLite3.ColType.Float) 
					    {
						    isFloat = true;
					    }
					    else if(columnType == SQLite3.ColType.Blob) 
					    {
						    isBlob = true;
					    }
					    else if(columnType == SQLite3.ColType.Null)
					    {
					    }


					    if (isString)
					    {
						
						    if (entityDescriptor != null)
						    {
							    EntityDescriptor.Attribute attribute = entityDescriptor.GetAttributeBasedOnColumnName(columnName);

							    if (attribute != null)
							    {

								    if (attribute.GetType().Equals(IDataTypeHandler.CSHARP_STRING_DATA_TYPE, StringComparison.OrdinalIgnoreCase))
								    {
									    tuple.Add(columnName, (String)columnValue);
								    }
								    else if (attribute.GetType().Equals(IDataTypeHandler.CSHARP_BOOLEAN_PRIMITIVE_DATA_TYPE, StringComparison.OrdinalIgnoreCase))
								    {
									    tuple.Add(columnName, columnValue.Equals(Boolean.TrueString.ToString()) ? true : false);
								    }
								    else if (attribute.GetType().Equals(IDataTypeHandler.CSHARP_BOOLEAN_DATA_TYPE, StringComparison.OrdinalIgnoreCase))
								    {
									    tuple.Add(columnName, columnValue.Equals(Boolean.TrueString.ToString()) ? Boolean.TrueString : Boolean.FalseString);
								    }
								    else if (attribute.GetType().Equals(IDataTypeHandler.JAVASCRIPT_STRING_DATA_TYPE, StringComparison.OrdinalIgnoreCase))
								    {
									    tuple.Add(columnName, (String)columnValue);
								    }
							    }
							    else
							    {
								    tuple.Add(columnName, columnValue);
							    }
						    }
						    else
						    {
							    tuple.Add(columnName, columnValue);
						    }
					    }
					    else if (isLong)
					    {

						    if (entityDescriptor != null)
						    {
							    EntityDescriptor.Attribute attribute = entityDescriptor.GetAttributeBasedOnColumnName(columnName);

							    if (attribute != null)
							    {
								    if (attribute.GetType().Equals(IDataTypeHandler.JAVASCRIPT_NUMBER_DATA_TYPE, StringComparison.OrdinalIgnoreCase))
								    {
									    tuple.Add(columnName, columnValue);
								    }
							    }
							    else
							    {
								    tuple.Add(columnName, columnValue);
							    }
						    }
						    else
						    {
							    tuple.Add(columnName, columnValue);
						    }
					    }
					    else if (isFloat)
					    {

						    if (entityDescriptor != null)
						    {
							    EntityDescriptor.Attribute attribute = entityDescriptor.GetAttributeBasedOnColumnName(columnName);

							    if (attribute != null)
							    {
								    if (attribute.GetType().Equals(IDataTypeHandler.JAVASCRIPT_NUMBER_DATA_TYPE, StringComparison.OrdinalIgnoreCase))
								    {
									    tuple.Add(columnName, columnValue);
								    }
							    }
							    else
							    {
								    tuple.Add(columnName, columnValue);
							    }
						    }
						    else
						    {
							    tuple.Add(columnName, columnValue);
						    }

						    tuple.Add(columnName, columnValue);
					    }
					    else if (isBlob)
					    {
						    tuple.Add(columnName, columnValue);
					    }
				    }

				    tuples.Add((Dictionary<String, Object>)tuple);
			    }

			    SQLite3.Finalize(statement);

			    return tuples.GetEnumerator();

            #elif WINDOWS
                foreach (SQLiteQueryRow cursor in statement)
                {

                    IDictionary<String, Object> tuple = new Dictionary<String, Object>();

                    List<SQLiteQueryColumn> columns = cursor.column;
                    if (columns == null || columns.Count <= 0)
                    {
                        continue;
                    }

                    foreach (SQLiteQueryColumn column in columns)
                    {

                        String columnName = column.Key;
                        Object columnValue = column.Value;


                        bool isString = false;
                        bool isLong = false;
                        bool isFloat = false;
                        bool isBlob = false;

                        if (columnValue != null)
                        {

                            if (columnValue.GetType().FullName.Equals(typeof(String).FullName, StringComparison.OrdinalIgnoreCase))
                            {
                                isString = true;
                            }
                            else if (columnValue.GetType().FullName.Equals(typeof(long).FullName, StringComparison.OrdinalIgnoreCase))
                            {
                                isLong = true;
                            }
                            else if (columnValue.GetType().FullName.Equals(typeof(float).FullName, StringComparison.OrdinalIgnoreCase))
                            {
                                isFloat = true;
                            }
                            else if (columnValue.GetType().FullName.Equals(typeof(byte).FullName, StringComparison.OrdinalIgnoreCase))
                            {
                                isBlob = true;
                            }
                        }
                        else
                        {
                            isString = true;
                        }


                        if (isString)
                        {

                            if (entityDescriptor != null)
                            {
                                EntityDescriptor.Attribute attribute = entityDescriptor.GetAttributeBasedOnColumnName(columnName);

                                if (attribute != null)
                                {

                                    if (attribute.GetType().Equals(IDataTypeHandler.CSHARP_STRING_DATA_TYPE, StringComparison.OrdinalIgnoreCase))
                                    {
                                        tuple.Add(columnName, (String)columnValue);
                                    }
                                    else if (attribute.GetType().Equals(IDataTypeHandler.CSHARP_BOOLEAN_PRIMITIVE_DATA_TYPE, StringComparison.OrdinalIgnoreCase))
                                    {
                                        tuple.Add(columnName, columnValue.Equals(Boolean.TrueString.ToString()) ? true : false);
                                    }
                                    else if (attribute.GetType().Equals(IDataTypeHandler.CSHARP_BOOLEAN_DATA_TYPE, StringComparison.OrdinalIgnoreCase))
                                    {
                                        tuple.Add(columnName, columnValue.Equals(Boolean.TrueString.ToString()) ? Boolean.TrueString : Boolean.FalseString);
                                    }
                                    else if (attribute.GetType().Equals(IDataTypeHandler.JAVASCRIPT_STRING_DATA_TYPE, StringComparison.OrdinalIgnoreCase))
                                    {
                                        tuple.Add(columnName, (String)columnValue);
                                    }
                                }
                                else
                                {
                                    tuple.Add(columnName, columnValue);
                                }
                            }
                            else
                            {
                                tuple.Add(columnName, columnValue);
                            }
                        }
                        else if (isLong)
                        {

                            if (entityDescriptor != null)
                            {
                                EntityDescriptor.Attribute attribute = entityDescriptor.GetAttributeBasedOnColumnName(columnName);

                                if (attribute != null)
                                {
                                    if (attribute.GetType().Equals(IDataTypeHandler.JAVASCRIPT_NUMBER_DATA_TYPE, StringComparison.OrdinalIgnoreCase))
                                    {
                                        tuple.Add(columnName, columnValue);
                                    }
                                }
                                else
                                {
                                    tuple.Add(columnName, columnValue);
                                }
                            }
                            else
                            {
                                tuple.Add(columnName, columnValue);
                            }
                        }
                        else if (isFloat)
                        {

                            if (entityDescriptor != null)
                            {
                                EntityDescriptor.Attribute attribute = entityDescriptor.GetAttributeBasedOnColumnName(columnName);

                                if (attribute != null)
                                {
                                    if (attribute.GetType().Equals(IDataTypeHandler.JAVASCRIPT_NUMBER_DATA_TYPE, StringComparison.OrdinalIgnoreCase))
                                    {
                                        tuple.Add(columnName, columnValue);
                                    }
                                }
                                else
                                {
                                    tuple.Add(columnName, columnValue);
                                }
                            }
                            else
                            {
                                tuple.Add(columnName, columnValue);
                            }

                            tuple.Add(columnName, columnValue);
                        }
                        else if (isBlob)
                        {
                            tuple.Add(columnName, columnValue);
                        }
                    }

                    tuples.Add((Dictionary<String, Object>)tuple);
                }


                return tuples.GetEnumerator();
            #endif

        }

        public void ExecuteMethod(String methodName, Object parameter)
        {
            MethodInfo method = null;

            try
            {
                method = sqliteDatabase.GetType().GetTypeInfo().GetDeclaredMethod(methodName);
            }
            catch (System.Exception exception)
            {
                Log.Log.Error(typeof(DatabaseImpl).FullName, "ExecuteMethod", "Exception caught while getting method, METHOD-NAME: " + methodName + ", " + exception.Message);
                throw new DatabaseException(typeof(DatabaseImpl).FullName, "ExecuteMethod", "Exception caught while getting method, METHOD-NAME: " + methodName + ", " + exception.Message);
            }

            try
            {
                method.Invoke(sqliteDatabase, new Object[] { });
            }
            catch (System.Exception exception)
            {
                Log.Log.Error(typeof(DatabaseImpl).FullName, "ExecuteMethod", "Exception caught while getting return value from method, METHOD-NAME: " + methodName + ", " + exception.Message);
                throw new DatabaseException(typeof(DatabaseImpl).FullName, "ExecuteMethod", "Exception caught while getting return value from method, METHOD-NAME: " + methodName + ", " + exception.Message);
            }
        }
    }
}
